Feature/myinfo

DokyungYou opened the pull request:
### 변경사항

[Chore: QueryDsl 의존성추가, 관련클래스 생성 및 연결 [aa85795] 부터 봐주시면 됩니다!
**AS-IS**
쿼리 dsl 추가
일부 코드 수정
서비스 테스트 코드 추가 

**TO-BE**
- [x] 테스트 코드
- [x] API 테스트

DokyungYou committed:
feat: 내 정보(태그 포함), 활동 보기 및 수정, 회원탈퇴 구현

DokyungYou committed:
Refactor: 필요없는 import문 삭제, Image엔티티, 레파지토리 생성

DokyungYou committed:
Merge branch 'develop' into feature/myinfo

# Conflicts:
#	solumon-backend/src/main/java/com/example/solumonbackend/member/controller/MemberController.java
#	solumon-backend/src/main/java/com/example/solumonbackend/member/entity/Member.java
#	solumon-backend/src/main/java/com/example/solumonbackend/member/repository/MemberRepository.java
#	solumon-backend/src/main/java/com/example/solumonbackend/member/service/MemberService.java

DokyungYou committed:
Style: develop과 병합 후 로드되지 않은 클래스 로드

DokyungYou committed:
Chore: QueryDsl 의존성추가, 관련클래스 생성 및 연결

DokyungYou committed:
!BREAKING CHANGE: 일부 엔티티 구조 변경

DokyungYou committed:
Refactor: 리뷰반영하여 코드수정

- Response<?> -> Response<DtoResponse> (타입을 명시하여 받도록 변경)
- @AuthenticationPrincipal MemberDetail 로 수정
- Request 클래스의 필드값의 패턴정규식 삭제
- Request 클래스의 스네이크법으로 작성했던 일부 필드 카멜케이스로 수정
- 풀,머지할때 정렬로 인해 다른 팀원 코드 건드렸던 것 원상복구
- 인자값으로 Entity 타입으로 넣었던것 pk값으로 변경
- 컨버팅 코드들을 스트림화
- Response 클래스 생성메소드 이름을 팀 계획에 맞게 수정
- 내 활동한 게시글목록 조회 기능 수정(nativeQuery -> QueryDsl + 페이징)

DokyungYou committed:
Comment: QueryDsl에 관련한 빌드설명 추가

DokyungYou committed:
Style: develop에서 새로 pull받은 내용과 병합

DokyungYou committed:
Refactor,Fix: 내 활동한 게시글목록 조회 기능 + 나머지기능(누락됐던 정규식, 어노테이션 추가)

- 일부 RequestDto 필드에 패턴정규식 추가
- MemberEntity reportCount 필드 삭제
- Dto클래스들에 @Getter 추가
- 내 활동한 게시글목록 조회 기능 관련 코드 리팩토링,  LocalDateTime 필드 바인딩문제 해결(관련 라이브러리 추가)

DokyungYou committed:
Refactor: 실수로 수정했던 부분 원복

DokyungYou committed:
Test: Member서비스 테스트 추가
Refactor: 내 정보 수정하기

ofgongmu commented on solumon-backend/src/main/java/com/example/solumonbackend/member/service/MemberService.java:
```diff
@@ -93,4 +118,148 @@ public GeneralSignInDto.Response signIn(GeneralSignInDto.Request request) {
+
+    // 닉네임 수정 시 사용중인 닉네임인지 확인
+    if (!request.getNickname().equals(member.getNickname())) {
+      if (memberRepository.existsByNickname(request.getNickname())) {
```
if문 &&으로 묶을 수 있을 것 같은데 혹시 분리하신 이유가 따로 있으실까요?

DokyungYou replied:
따로 큰 이유는 없고 저렇게 쓰는게 습관이 되었나봅니다.
앞으로 묶을 수 있는건 묶도록 의식하고 짜야겠어요

ofgongmu commented on solumon-backend/src/main/java/com/example/solumonbackend/member/service/MemberService.java:
```diff
@@ -93,4 +118,148 @@ public GeneralSignInDto.Response signIn(GeneralSignInDto.Request request) {
+    // 비밀번호 수정 시 새 비밀번호1 == 새 비밀번호2 인지 확인
+    if (request.getNewPassword1() != null) {
+      if (!request.getNewPassword1().equals(request.getNewPassword2())) {
+        throw new MemberException(NOT_CORRECT_NEW_PASSWORD);
```
이 경우 에러코드를 NEW_PASSWORDS_DO_NOT_MATCH 정도로 하면 문제가 무엇인지 조금 더 명확해질 것 같습니다!

DokyungYou replied:
확실히 더 명확한 것 같습니다!

ofgongmu commented on solumon-backend/src/main/java/com/example/solumonbackend/post/repository/PostRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,182 @@
+      return qpost.endAt.isNull().or(qpost.endAt.before(LocalDateTime.now()));
+    }
+
+    return null;
```
이 부분 elseIf로 묶어주시면 좋을 것 같아요! if Completed / else if Ongoing / else 에서 exception을 던져주는 방식으로요. 그럴 일은 없을 것 같긴 한데 null을 돌려줄 수도 있다는 일말의 가능성 때문에... 아예 차단하는 게 어떨까 싶습니다

DokyungYou replied:
원천봉쇄 좋습니다. 수정하겠습니다!

DokyungYou replied:
이런식으로도 수정해봤는데요. 어떻게 생각하세요??


  private OrderSpecifier<?> createOrderSpecifier(PostOrder order, QPost qpost) {

    try {
      if (qpost == null || order == null) {
        throw new NullPointerException("PostOrder 의 조건값이 없습니다.");
      }
      // 최신순)  post.createdAt.desc()
      if (order == PostOrder.POST_ORDER) {
        return qpost.createdAt.desc(); // OrderSpecifier<LocalDateTime>
      }
      // 투표참여인원)  post.voteCount.desc()
      if (order == PostOrder.MOST_VOTES) {
        return qpost.voteCount.desc(); // OrderSpecifier<Integer>
      }

      // 채팅참여인원)  post.chatCount.desc()
      if (order == PostOrder.MOST_CHAT_PARTICIPANTS) {
        return qpost.chatCount.desc();  // OrderSpecifier<Integer>
      }

      // null 이 아니고 위의 조건에도 걸리지 않았다면 잘못된 값이 들어왔다는 의미(그럴 확률은 거의 없겠지만..?)
      throw new IllegalArgumentException();

    } catch (NullPointerException | IllegalArgumentException e) {
      // 이부분은 처리를 어떤 방식으로 하는게 좋을지..
    }

    return null;
  }

ofgongmu replied:
이렇게도 괜찮을 것 같습니다! catch 부분은... 만일 발생한다면 클라이언트 측 문제라기보다는 서버 측 문제 같아서 500 에러를 뜨게 해야 할 것 같은데 이걸 저도 발생시켜본 적이 없어서 잘 모르겠네요... 일단은 PostException으로 처리해두고 나중에 exception status 고칠 때 다시 살펴보는 게 좋을 것 같습니다!

DokyungYou replied:
넵!

ofgongmu commented on solumon-backend/src/test/java/com/example/solumonbackend/member/service/MemberServiceTest2.java:
```diff
@@ -0,0 +1,509 @@
+  public void testGetMyInfo() {
+
+    // Given
+    Member fakeMember = MemberTestData.fakeMemberBuilder().build();
```
이 부분은 별도의 클래스를 만들기보다는 메소드를 테스트 시작 전마다 실행시켜주는 @BeforeEach라는 어노테이션으로도 처리할 수 있을 것 같아요. 쓰지 않는 정보도 있기는 하지만 그 편이 조금 더 깔끔하지 않을까...? 싶은데 어떻게 생각하세요?

DokyungYou replied:
말씀하신대로 수정했는데 전보다 깔끔해지고 코드가 확 줄어드네요!

ofgongmu commented on solumon-backend/src/main/java/com/example/solumonbackend/member/model/MemberInterestDto.java:
```diff
@@ -0,0 +1,44 @@
+  @AllArgsConstructor
+  public static class Request {
+
+    @NotNull
```
관심사 설정을 하지 않아도 되는 걸로 기억하는데 이 부분 @NotNull이어도 괜찮을까요?

DokyungYou replied:
앗 그랬죠. 메소드도 수정해야겠네요.

DokyungYou replied:
수정하면서 생각났는데
어차피 List로 받는거니까 선택을 안하면 깡통 List가 오는거 아닌가?
라고 생각하면서 @NotNull을 붙였었네요. 
속에 값이 없더라도 초기화된 List를 받아오는거 아닌가? 하고 생각했었어요.

ofgongmu commented on solumon-backend/src/main/java/com/example/solumonbackend/post/repository/PostRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,182 @@
+    List<MyParticipatePostDto> resultContents
+        = jpaQueryFactory.select(Projections.bean(MyParticipatePostDto.class,
+            qpost.postId,
+            qpost.member.nickname.as("writerNickname"), // 매칭되는 이름이 다를 땐 별칭사용
```
혹시 이 부분이 어떤 뜻인지 조금만 더 설명해주실 수 있으실까요?

DokyungYou replied:
as("writerNickname")를 붙이지 않았을 경우 저 데이터를 받아야하는 Dto클래스의 필드는 String nickname 이어야만 합니다.
그러나  nickname보단 보자마자 해당 글을 작성한 멤버의 닉네임으로 인식할 수 있게끔 writerNickname으로 선언해준 상태고 

writerNickname에 nickname 컬럼의 값이 바인딩되려면 as("writerNickname")을 넣어줘야합니다

ofgongmu reviewed (commented):
리뷰 남겼습니다. querydsl 공부하시느라 고생 너무 많으셨습니다...!

Gayeon-H commented on solumon-backend/src/main/java/com/example/solumonbackend/member/controller/MemberController.java:
```diff
@@ -56,4 +68,92 @@ public void exception() throws RuntimeException {
+  public ResponseEntity<MemberLogDto.Info> getMyInfo(@AuthenticationPrincipal MemberDetail memberDetail) {
+
+    MemberLogDto.Info response = memberService.getMyInfo(memberDetail.getMember());
+    return ResponseEntity.ok().body(response);
```
response가 한번만 쓰인다면 아예 body 안에 써도 괜찮지 않을까요? 따로 빼두신 이유는 가독성 때문일까요?
개인 프로젝트를 할 때 지역변수가 한번만 쓰인다면 안 써도 되는 메모리 공간을 쓰기 때문에 만들지 말고 바로 넣으라는 멘토님 말씀을 들은 적이 있어서요.

DokyungYou replied:
여러 번 쓰일 게 아니라면 지역변수로 따로 빼지 않는게 좋을것이라는 것은 인지를 못하고 있던 건 아니었는데
들었던 강의(100제)에서 매번 저렇게 썻던게 습관이 돼서  큰 생각 없이 전부 저렇게 짯네요..

Gayeon-H commented on solumon-backend/src/main/java/com/example/solumonbackend/post/entity/Post.java:
```diff
@@ -37,6 +37,8 @@ public class Post {
   @LastModifiedDate
   private LocalDateTime modifiedAt;
 
+  private String thumbnailUrl;
```
이거는 제가 없을 때 추가된 부분이군요! 프론트에서 대표이미지를 어떻게 표시해서 넘겨줄지는 아직 논의하지 않은 상태죠??

DokyungYou replied:
넵, 아직 그 부분에 대해서는 논의가 안된 상태고, 
저 코드의 필드는 일단 erd 수정해주신 부분을 반영해서 만들었습니다

Gayeon-H reviewed (commented):
queryDSL 설정하고 코드 짜신다고 수고 많으셨어요!!

GyeongSe99 commented on solumon-backend/src/main/java/com/example/solumonbackend/member/controller/MemberController.java:
```diff
@@ -1,5 +1,124 @@
+   * @return
+   */
+  @GetMapping("/users/mylog")
+  public ResponseEntity<?> getMyActivePosts(
```
와일드 카드 말고 실제 리턴값으로 바꿔 주세용

DokyungYou replied:
앗 이건 멘토링 전에 작성됐던 코드에요.
다시 리퀘스트 여니까 저번것까지 같이 떠서요..
Chore 커밋 부터 봐주시면 됩니당

GyeongSe99 replied:
아항 넵 알겠습니다

GyeongSe99 commented on solumon-backend/src/main/java/com/example/solumonbackend/member/entity/Ban.java:
```diff
@@ -6,8 +6,16 @@
+import lombok.NoArgsConstructor;
 
 @Entity
+@Data
```
builder가 있어서 getter만 있어도 될 것 같은데 @Data쓰신 이유가 따로 있으신가요?

DokyungYou replied:
푸시 뒤에 확인했었습니다.
지금은 수정된 상태입니다!

GyeongSe99 reviewed (commented):
리뷰 남겼습니다

DokyungYou committed:
Refactor,Test: 피드백 반영하여 수정

DokyungYou committed:
Test: MemberControllerTest2
- 내 정보 조회 (o)
- 내 활동한 게시글목록 조회(보류)
- 내 정보 수정 (o)
- 회원 탈퇴 (o)
- 관심 주제 선택 (o)

DokyungYou committed:
Refactor: 지역변수로 받지않고 바로 리턴하도록 수정

DokyungYou committed:
Merge branch 'develop' into feature/myinfo

# Conflicts:
#	solumon-backend/src/main/java/com/example/solumonbackend/global/exception/ErrorCode.java
#	solumon-backend/src/main/java/com/example/solumonbackend/member/controller/MemberController.java
#	solumon-backend/src/main/java/com/example/solumonbackend/member/model/GeneralSignUpDto.java
#	solumon-backend/src/main/java/com/example/solumonbackend/member/repository/MemberRepository.java
#	solumon-backend/src/main/java/com/example/solumonbackend/member/service/MemberService.java
#	solumon-backend/src/test/java/com/example/solumonbackend/member/controller/MemberControllerTest2.java

DokyungYou committed:
Style: develop브랜치 pull, feature/myinfo 에 merge

DokyungYou committed:
Style: develop브랜치 pull, feature/myinfo 에 merge

DokyungYou committed:
Test,Refactor: 컨트롤러 테스트 코드 수정

- MemberServiceTest3 -> MemberServiceTest_MyInfo 로 변경
- 테스트코드 수정
- 내 활동 가져오기 쿼리문 잘못된 조건 수정 (마감중, 진행중이 반대로 돼있었음)

DokyungYou commented:
컨트롤러 테스트 코드 정상작동까지 마쳤습니다

DokyungYou committed:
Refactor:  enum 이름 변경

DokyungYou committed:
Test:  TestMyParticipatePostData 레이아웃 수정

ofgongmu reviewed (approved):
개인적으로는 빌더는 필드가 많더라고 연속적으로 나열하는 것보다는 하나씩 아래로 띄워서 하는 게 보기가 더 좋은 것 같아요! 어떤 필드가 있고 그 필드에 뭐가 들어가는지 볼 수 있는 게 빌더의 장점이기도 하고요 ㅎㅎ 고생 많으셨습니다!

DokyungYou committed:
Test:  TestMyParticipatePostData 레이아웃 수정

- 빌더의 형태를 하나씩 아래로 띄우는 방식으로 수정

Gayeon-H reviewed (approved):
저는 merge하셔도 괜찮을 것 같아요. 수고많으셨습니다!

DokyungYou commented:
머지했습니다!
